---
title: "MCF10A_timecourse_RNAseq"
author: "Haley Greenyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, echo=FALSE, warning=FALSE, message=FALSE}
library(tximport)
library(GenomicFeatures)
library(readr)
library(gprofiler2)
library(DESeq2)
library(tidyverse)
library(hciR)
library(pheatmap)
library(vidger)
library(ggVennDiagram)
library(UpSetR)
library(ggdendro)
library(geneplotter)
library(kableExtra)
library(openxlsx)
library(ensembldb)
library(dplyr)
library(ggplot2)
library(tibble)
library(magrittr)
#library(RNAseqQC)

library(dplyr)

library(EnhancedVolcano)
library(ggVennDiagram)

library(hciR)

source("/slipstream_old/home/hgreenyer/Rscripts/RNAseq_DEA_functions.R")
```
# MCF10A FKBP Timecourse RNAseq
## RNA-seq QC all samples
```{r directories, include=FALSE}

#reference anno
ref_dir <- "/slipstream/home/hgreenyer/gencode.v36.annotation.gtf"
#salmon quant files
s_dir <- "/slipstream_old/home/hgreenyer/PROJECTS/MCF10A_RUNX1_TIMECOURSE_2023/MCF10A_TimeCourse_RNAseq/MCF10A_FKBP_DATA_ALL/PIPELINE_OUT/"
#output directories
fig_out_dir <- "/slipstream_old/home/hgreenyer/PROJECTS/MCF10A_RUNX1_TIMECOURSE_2023/MCF10A_TimeCourse_RNAseq/Figures/"
file_out_dir <- "/slipstream_old/home/hgreenyer/PROJECTS/MCF10A_RUNX1_TIMECOURSE_2023/MCF10A_TimeCourse_RNAseq/DEA_output/"
out_dir <- "/slipstream_old/home/hgreenyer/PROJECTS/MCF10A_RUNX1_TIMECOURSE_2023/MCF10A_TimeCourse_RNAseq/DEA_output/"
```

```{r generate annotation, warning=FALSE, message=FALSE, echo = FALSE}

#make annotation from sample directory 
anno_all <- get_anno(s_dir)
```

#Generation Annotation 
#
```{r edit annotation}

anno_all$replicate <- as.character(anno_all$replicate)

#convert factors to characters
anno_all$cell <- as.character(anno_all$cell)
anno_all$condition <- as.character(anno_all$condition)
anno_all$timepoint <- as.character(anno_all$timepoint)

#fixes 
#inconsistent replicate naming
anno_all$replicate[anno_all$replicate == "1"] <- "R1"
anno_all$replicate[anno_all$replicate == "2"] <- "R2"
anno_all$replicate[anno_all$replicate == "3"] <- "R3"

#inconsistent cell line naming 
anno_all$cell[anno_all$cell == "FKBP" | anno_all$cell == 'MCF10A'] <- 'C7'

#inconsistent time naming 
anno_all$timepoint[anno_all$timepoint == "48"] <- '48hr'
anno_all$timepoint[anno_all$timepoint == "3H"] <- '3hr'
anno_all$timepoint[anno_all$timepoint == "6D"] <- '6d'


#downsampled files were in different order - swap cols
d_rows <- !grepl("MCF10A", anno_all$sample)
temp <- anno_all$condition[d_rows]
anno_all$condition[d_rows] <- anno_all$timepoint[d_rows]
anno_all$timepoint[d_rows] <- temp

#inconsistent condition naming 
anno_all$condition[anno_all$condition == "DTAG"] <- 'dTAG'


#condense sample names 
get_name <- function(x){
  parts <- unlist(strsplit(x, "_"))
  name <- paste(parts[1:4], collapse = "_")
  return(name)
}

anno_all$sample <- sapply(as.character(anno_all$sample), get_name)
anno_all$experiment <- as.factor(c(rep(3,18),rep(2,3),rep(1,3),rep(2,6),rep(1,3),rep(2,3),rep(3,12)))

write.csv(x=anno_all, file=paste(out_dir,'MCF10A_timecourse_RNAseq_all_anno.csv',sep=''))
```

```{r subset annotation}

#subset per cell line 
anno_parent <- anno_all %>% dplyr::filter(anno_all[['cell']] == 'Parent')
anno_C7 <- anno_all %>% dplyr::filter(anno_all[['cell']] == 'C7')

#look at DMSO only for Parental vs C7 comparison 
anno_DMSO <- anno_all %>% dplyr::filter(anno_all[['condition']] == 'DMSO' & anno_all[['timepoint']] != '12hr')

#subset by experiment for PCA plot 
anno_E1 <- anno_all %>% dplyr::filter(anno_all[['experiment']] != 3)
anno_E2 <- anno_all %>% dplyr::filter(anno_all[['experiment']] == 3)

```

#Generating Raw Counts
#
```{r generate raw counts (all)}
txi_all <- get_txi(ref_dir, s_dir, anno_all)

#get DESeq object 
dds_all <- get_dds(anno_all, txi_all)

#export raw count table 
export_counts(dds = dds_all, norm = FALSE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_all')

```

```{r generate DESeq objects for subsets}
#generate dds objects for each subset and export raw counts where appropriate 

#parental 10A
txi_parent <- get_txi(ref_dir, s_dir, anno_parent)
dds_parent <- get_dds(anno_parent, txi_parent)
vst_parent <- varianceStabilizingTransformation(dds_parent, blind=TRUE)
export_counts(dds = dds_parent, norm = FALSE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_Parental')

#C7 10A
txi_C7 <- get_txi(ref_dir, s_dir, anno_C7)
dds_C7 <- get_dds(anno_C7, txi_C7)
vst_C7 <- varianceStabilizingTransformation(dds_C7, blind=TRUE)
export_counts(dds = dds_C7, norm = FALSE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_C7')

#first experimental set 
txi_E1 <- get_txi(ref_dir, s_dir, anno_E1)
dds_E1 <- get_dds(anno_E1, txi_E1)
vst_E1 <- varianceStabilizingTransformation(dds_E1, blind=TRUE)
#second experimental set 
txi_E2 <- get_txi(ref_dir, s_dir, anno_E2)
dds_E2 <- get_dds(anno_E2, txi_E2)
vst_E2 <- varianceStabilizingTransformation(dds_E2, blind=TRUE)

```


```{r dds object DMSO samples}
#DMSO only for checking parental vs C7
txi_DMSO <- get_txi(ref_dir, s_dir, anno_DMSO)

coldata = anno_DMSO
rownames(coldata) = anno_DMSO$sample

#special case because condition is not relevant here 
dds_DMSO <- DESeqDataSetFromTximport(txi  = txi_DMSO, 
                                    colData = coldata, 
                                    design = ~replicate + cell)

vst_DMSO <- varianceStabilizingTransformation(dds_DMSO, blind=TRUE)

```

#Filtering and Normalization 
#
```{r filtering and normalization all samples, warning=FALSE, message=FALSE, echo = FALSE}
#filter all samples together 
#remove low abundance features 
keep <- rowSums(counts(dds_all) >= 10) >= 3

dds_all <- dds_all[keep ,]

#export normalized count table 
export_counts(dds = dds_all, norm = TRUE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_all')

#vst objects 
vst_all <- varianceStabilizingTransformation(dds_all, blind=TRUE)
vst_all_mat <- assay(vst_all)

```

```{r filtering and normalization parental, warning=FALSE, message=FALSE, echo = FALSE}
#repeat for all subsets 
#parental 10A
keep <- rowSums(counts(dds_parent) >= 10) >= 3
dds_parent <- dds_parent[keep ,]
export_counts(dds = dds_parent, norm = TRUE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_parental')
vst_parent <- varianceStabilizingTransformation(dds_parent, blind=TRUE)

#C7 10A
keep <- rowSums(counts(dds_C7) >= 10) >= 3
dds_C7 <- dds_C7[keep ,]
export_counts(dds = dds_C7, norm = TRUE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_C7')
vst_C7 <- varianceStabilizingTransformation(dds_C7, blind=TRUE)

#Experiment 1
keep <- rowSums(counts(dds_E1) >= 10) >= 3
dds_E1 <- dds_E1[keep ,]
vst_E1 <- varianceStabilizingTransformation(dds_E1, blind=TRUE)

#Experiment 2 
keep <- rowSums(counts(dds_E2) >= 10) >= 3
dds_E2 <- dds_E2[keep ,]
vst_E2 <- varianceStabilizingTransformation(dds_E2, blind=TRUE)

#DMSO only 
keep <- rowSums(counts(dds_DMSO) >= 10) >= 3
dds_DMSO <- dds_DMSO[keep ,]
vst_DMSO <- varianceStabilizingTransformation(dds_DMSO, blind=TRUE)

```

# Quality Control Figures
## library sizes and gene count densities
```{r count densities, message=FALSE, warning=FALSE}
#SKIP RUN 
bulkQC_figs(dds_all, txi_all, anno_all)

```

```{r batch correction}
#PCA plots for all samples 

#all timepoints together
pcaData <- plotPCA(vst_all, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 
#default
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

#by cell
ggplot(pcaData, aes(x=PC1, y=PC2, color=cell, shape=condition)) + 
  geom_point(size =3, aes(fill=cell)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

#by experiment
ggplot(pcaData, aes(x=PC1, y=PC2, color=experiment, shape=condition)) + 
  geom_point(size =3, aes(fill=experiment)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  


```

```{r PCA C7 correction}
#PCA plots for all samples 

#all timepoints together
pcaData <- plotPCA(vst_C7, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 
#default
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

#by experiment
ggplot(pcaData, aes(x=PC1, y=PC2, color=experiment, shape=condition)) + 
  geom_point(size =3, aes(fill=experiment)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

```

```{r PCA E1 and E2 correction}
#PCA plots for all samples 

#all timepoints together
pcaData <- plotPCA(vst_E1, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 
#default
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

pcaData <- plotPCA(vst_E2, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 
#default
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  
```

```{r PCA Parental correction}
#PCA plots for all samples 

#all timepoints together
pcaData <- plotPCA(vst_parent, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 
#default
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  

```


## Differential Expression
### DMSO vs DTAG at 3H, 48H, 6D 
```{r subset dds for each timepoint, message=FALSE, warning=FALSE, include=FALSE}

#get DESeq objects for each timepoint 
#C7 samples only 

dds_3H <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '3hr')

dds_6H <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '6hr')

dds_12H <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '12hr')

dds_24H <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '24hr')

dds_48H <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '48hr')

dds_6D <- subset_dds(txi.salmon = txi_C7, annotation = anno_C7, sub_col = 'timepoint', sub = '6d')

#parental samples
dds_P_6H <- subset_dds(txi.salmon = txi_parent, annotation = anno_parent, sub_col = 'timepoint', sub = '6hr')

dds_P_48H <- subset_dds(txi.salmon = txi_parent, annotation = anno_parent, sub_col = 'timepoint', sub = '48')

```

```{r get DMSO vs DTAG DEGs for each timepoint, message=FALSE, warning=FALSE, include=FALSE}

#get DEG lists DMSO vs DTAG for each timepoint 

#returns degs and expression - exluding samples without gene names 
#previously used very stringent oadj cutoff of 0.01 and LFC of 0.58, changing to 0.05 and increasing LFC cutoff to 1

deg_3H <- get_DEGs(dds = dds_3H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_3H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_6H <- get_DEGs(dds = dds_6H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_6H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_12H <- get_DEGs(dds = dds_12H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_12H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_24H <- get_DEGs(dds = dds_24H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_24H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_48H <- get_DEGs(dds = dds_48H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_48H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_6D <- get_DEGs(dds = dds_6D, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7_DMSOvDTAG_6D',l2fc_cutoff = 1, padj_cutoff = 0.05)


#parental
deg_P_6H <- get_DEGs(dds = dds_P_6H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_Parental_DMSOvDTAG_6H',l2fc_cutoff = 1, padj_cutoff = 0.05)

deg_P_48H <- get_DEGs(dds = dds_P_48H, ctrst = c('condition','dTAG','DMSO'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_Parental_DMSOvDTAG_6H',l2fc_cutoff = 1, padj_cutoff = 0.05)


#DMSO only, parental vs C7
tmp <- DESeq(dds_DMSO)
  
deg_PC <- get_DEGs(dds = tmp, ctrst = c('cell','C7','Parent'), export_degs = TRUE, out_dir = file_out_dir, out_pref = 'MCF10a_C7vParental',l2fc_cutoff = 1, padj_cutoff = 0.05)


```
