---
title: "MCF10A_timecourse_RNAseq"
author: "Haley Greenyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, echo=FALSE, warning=FALSE, message=FALSE}
library(tximport)
library(GenomicFeatures)
library(readr)
library(gprofiler2)
library(DESeq2)
library(tidyverse)
library(pheatmap)
library(ggVennDiagram)
library(UpSetR)
library(openxlsx)
library(ensembldb)
library(dplyr)
library(ggplot2)
library(tibble)
library(magrittr)
#library(RNAseqQC)
library(EnhancedVolcano)

source("~/RNAseq_DEA_functions.R")
```
# RMA-seq Differential Expression Analysis
## RNA-seq QC all samples
```{r directories, include=FALSE}
#input directories
ref_dir <- "/path/to/annotation.gtf"
s_dir <- "path/to/salmon_quant/directories/"

#output directories
fig_out_dir <- "/path/for/Figures/"
out_dir <- "/path/for/DEA_output_files/"
```
#Generate Annotation 
```{r generate annotation, warning=FALSE, message=FALSE, echo = FALSE}
anno_all <- get_anno(s_dir)
```

```{r export annotation}
#edit annotation here as needed 
write.csv(x=anno_all, file=paste(out_dir,'MCF10A_timecourse_RNAseq_all_anno.csv',sep=''))
```

```{r subset annotation}
#subset annotation here if needed by whatever groupings will be used for DEA
```

#Generating Raw Counts
```{r generate raw counts (all)}
txi_all <- get_txi(ref_dir, s_dir, anno_all)

#get DESeq object 
dds_all <- get_dds(anno_all, txi_all)

#export raw count table 
export_counts(dds = dds_all, norm = FALSE, out_dir = file_out_dir, out_pref = 'Output_Prefix')
```

```{r generate DESeq objects for subsets}
#if you need to segment your data for DEA, generate subset DESeq objects 
```

#Filtering and Normalization 
```{r filtering and normalization, warning=FALSE, message=FALSE, echo = FALSE}
#remove low abundance features 
keep <- rowSums(counts(dds_all) >= 10) >= 3

dds_all <- dds_all[keep ,]

#export normalized count table 
export_counts(dds = dds_all, norm = TRUE, out_dir = file_out_dir, out_pref = 'MCF10A_RUNX1_TimeCourse_all')

#vst objects 
vst_all <- varianceStabilizingTransformation(dds_all, blind=TRUE)
vst_all_mat <- assay(vst_all)

#repeat this step for any subset onbjects that DEA will be performed on 
```

# Quality Control Figures
## library sizes and gene count densities
```{r count densities, message=FALSE, warning=FALSE}
#SKIP RUN WHEN NOT NECESSARY, CAN BE SLOW
bulkQC_figs(dds_all, txi_all, anno_all)
```
#PCA Plots 
```{r batch correction} 
#can use limma to correct batch effects here as needed
pcaData <- plotPCA(vst_all, intgroup=c('condition','timepoint','replicate','cell','experiment'), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar")) 

#default, no corrections
ggplot(pcaData, aes(x=PC1, y=PC2, color=timepoint, shape=condition)) + 
  geom_point(size =3, aes(fill=timepoint)) + 
  ggtitle('MCF10a RUNX1 Degron PCA') +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance"))  
```

## Differential Expression
```{r differential tests, message=FALSE, warning=FALSE, include=FALSE}
#example call
#returns degs and expression - exluding samples without gene names 
#note that C2 is the reference/control condition here
deg_all <- get_DEGs(dds = dds_all, ctrst = c('condition','C1','C2'), export_degs = TRUE, out_dir = out_dir, out_pref = 'Out_DEG_prefix',l2fc_cutoff = 1, padj_cutoff = 0.05)
```
